protoc-gen-gogo/generator/generator.go

```diff
@@ -3209,6 +3217,106 @@ func (g *Generator) generateCommonMethods(mc *msgCtx) {
+	g.P("func (m *", typeName, ") Clone() ", returnName, "{")
+	g.In()
+	g.P("if m == nil {")
+	g.In()
+	g.P("return nil")
+	g.endBlock()
+	g.P("cloned := new(", typeName, ")")
+	g.P("*cloned = *m")
+	g.P()
+}
+
+func (g *Generator) cloneGenericFooter() {
+	g.P("return cloned")
+	g.endBlock()
+}
+
+func (g *Generator) ifNotNilHeader(name string) {
+	g.P("if m.", name, "!= nil {")
+	g.In()
+}
+
+func (g *Generator) endBlock() {
+	g.Out()
+	g.P("}")
+}
+
+func (g *Generator) generateClone(mc *msgCtx, topLevelFields []topLevelField, mapFields map[string]*GoMapDescriptor) {
+	g.cloneGenericHeader(mc.goName, fmt.Sprintf("*%s", mc.goName))
+
+	for _, field := range topLevelFields {
+		switch typedField := field.(type) {
+		case *simpleField:
+			switch *typedField.protoField.Type {
+			case descriptor.FieldDescriptorProto_TYPE_MESSAGE:
+				// If repeated do x, otherwise do y
+				if *typedField.protoField.Label == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+					if mapDesc, ok := mapFields[typedField.goName]; ok {
+						// Handle Map
+						g.ifNotNilHeader(typedField.goName)
+						g.P("cloned.", typedField.goName, "= make(", mapDesc.GoType, ", ", "len(m.", typedField.goName, ")", ")")
+						g.P("for k, v := range m.", typedField.goName, " {")
+						g.In()
+						if *mapDesc.ValueField.Type == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
+							g.P("cloned.", typedField.goName, "[k] = v.Clone()")
+						} else {
+							g.P("cloned.", typedField.goName, "[k] = v")
+						}
+						g.endBlock()
+						g.endBlock()
+					} else {
+						// Handle Slice
+						g.ifNotNilHeader(typedField.goName)
+						g.P("cloned.", typedField.goName, "= make(", typedField.goType, ", len(m.", typedField.goName, ")", ")")
+						g.P("for idx, v := range m.", typedField.goName, " {")
+						g.In()
+						// All slices here are messages, otherwise, they are not of type message
+						g.P("cloned.", typedField.goName, "[idx] = v.Clone()")
+						g.endBlock()
+						g.endBlock()
+					}
+				} else {
+					g.ifNotNilHeader(typedField.goName)
+					switch typedField.goType {
+					case "*types.Timestamp":
+						g.P("cloned.", typedField.goName, "= new(types.Timestamp)")
+						g.P("cloned.", typedField.goName, ".Seconds = m.", typedField.goName, ".Seconds")
+						g.P("cloned.", typedField.goName, ".Nanos = m.", typedField.goName, ".Nanos")
+					case "*types.Any":
+						g.P("cloned.", typedField.goName, "= new(types.Any)")
+						g.P("cloned.", typedField.goName, ".TypeUrl = m.", typedField.goName, ".TypeUrl")
+						g.P("cloned.", typedField.goName, ".Value = make([]byte, len(m.", typedField.goName, ".Value))")
+						g.P("copy(cloned.", typedField.goName, ".Value", ", m.", typedField.goName, ".Value)")
+					default:
+						g.P("cloned.", typedField.goName, "=", "m.", typedField.goName, ".Clone()")
+					}
+					g.endBlock()
+				}
+			default:
+				if *typedField.protoField.Label == descriptor.FieldDescriptorProto_LABEL_REPEATED {
```

Comment:
You need to special-case the `bytes` type, which is an (otherwise aliasing) slice even if it's not repeated.

Reply:
Thinking about it, you need to special-case `bytes` everywhere - for repeated fields, and also for `map` values